chaincode/contracts/AssetManager.sol

```diff
@@ -0,0 +1,190 @@
+        if (amount == 0){ amount = balance; }
+        // check msg.senders balance
+        if (amount > balance) {
+            emit WithdrawalFailed(
+                msg.sender,
+                amount,
+                balance,
+                "Insufficient Locked Funds to Withdraw "
+            );
+            return;
+        }
+
+        require(amount <= balance);
+        // withdraw funds from the contract (update userBalance before transfer to protect from reentracy attack)
+        uint256 newBalance = balance - amount;
+        userBalances[address(msg.sender)] = newBalance;
+        payable(msg.sender).transfer(amount);
+
+        // update the authorized amount.
+        uint256 currentLimit = getUserAuthorization(msg.sender);
+        if (currentLimit > newBalance ) { authorize(newBalance); }
+
+        // update the userBalance
+        emit WithdrawalSuccesful(
+            msg.sender,
+            amount,
+            userBalances[address(msg.sender)]
+        );
+    }
+
+    function authorize(uint256 limit) public {
+        // check msg.senders balance
+        uint256 balance = getUserBalance(msg.sender);
+        // uint256 currentLimit = getUserAuthorization(msg.sender);
+        // uint256 newLimit = currentLimit + limit;
+        if (balance < limit) {
+            emit AuthorizationFailed(
+                msg.sender,
+                balance,
+                limit,
+                "New limt amount greater than users balance"
+            );
+            return;
+        }
+
+        require(balance >= limit);
+        userAuthorizations[address(msg.sender)] = limit;
+
+        // update the userBalance
+        emit AuthorizationSuccesful(
+            msg.sender,
+            limit,
+            balance
+        );
+    }
+    function send(uint256 amount, address from, address to) public onlyOwner() {
+        uint256 balance =  userBalances[from];
+        uint256 limit = userAuthorizations[from];
+        require (amount != 0);
+        // check from balance
+        if (amount > balance) {
+            emit SendFailed(
+                from,
+                to,
+                amount,
+                balance,
+                limit,
+                "Insufficient Locked Funds to Send "
+            );
+            return;
+        }
+         require(amount <= balance);
+        // check from balance
+        if (amount > limit) {
+            emit SendFailed(
+                from,
+                to,
+                amount,
+                balance,
+                limit,
+                "Insufficient Authorized Funds to Send "
+            );
+            return;
+        }
+         require(amount <= limit);
+        // withdraw funds from the contract (update userBalance before transfer to protect from reentracy attack)
+        uint256 newBalance = balance - amount;
+        userBalances[address(from)] = newBalance;
+        payable(to).transfer(amount);
```

Comment:
Potential re-entrance attack here. All adjustments (including authorizations) should be made prior to transfer

Reply:
moved transfer after authorize

Reply:
updated